kolibri/content/utils/annotation.py

```diff
@@ -1,38 +1,163 @@
-from ..models import ChannelMetadata, ChannelMetadataCache
 from .channels import get_channel_ids_for_content_database_dir
+from .paths import get_content_file_name, get_content_storage_file_path
+from .sqlalchemybridge import Bridge
+
+logging = logger.getLogger(__name__)
 
+CONTENT_APP_NAME = KolibriContentConfig.label
 
 def update_channel_metadata_cache():
     """
-    After a channel is imported, or when the devserver is started,
+    If we are potentially moving from a version of Kolibri that did not import its content data,
     scan through the settings.CONTENT_DATABASE_DIR folder for all channel content databases,
     and pull the data from each database's ChannelMetadata object to update the ChannelMetadataCache
     object in the default database to ensure they are in sync.
     """
-    db_names = get_channel_ids_for_content_database_dir(settings.CONTENT_DATABASE_DIR)
-    # Delete ChannelMetadataCache objects in default db that are not found in CONTENT_DATABASE_DIR
-    ChannelMetadataCache.objects.exclude(id__in=db_names).delete()
-    # sync the ChannelMetadataCache objects in default db with ChannelMetadata objects in CONTENT_DATABASE_DIR
-    for db_name in db_names:
-        with using_content_database(db_name):
-            update_values = ChannelMetadata.objects.values()[0]
-
-        ch_metadata_obj, _ = ChannelMetadataCache.objects.update_or_create(
-            id=db_name,
-            defaults=update_values,
+    from .channel_import import import_channel_from_local_db
+    channel_ids = get_channel_ids_for_content_database_dir(settings.CONTENT_DATABASE_DIR)
+    for channel_id in channel_ids:
+        if not ChannelMetadata.objects.filter(id=channel_id).exists():
+            import_channel_from_local_db(channel_id)
+
+def set_leaf_node_availability_from_local_file_availability():
+    bridge = Bridge(app_name=CONTENT_APP_NAME)
+
+    ContentNodeTable = bridge.get_table(ContentNode)
+    FileTable = bridge.get_table(File)
+    LocalFileTable = bridge.get_table(LocalFile)
+
+    connection = bridge.get_connection()
+
+    file_statement = select([LocalFileTable.c.available]).where(
+        and_(
+            FileTable.c.local_file_id == LocalFileTable.c.id,
+        )
+    ).limit(1)
+
+    logging.info('Setting availability of File objects based on LocalFile availability')
+
+    connection.execute(FileTable.update().values(available=file_statement).execution_options(autocommit=True))
+
+    contentnode_statement = select([FileTable.c.contentnode_id]).where(
+        and_(
+            FileTable.c.available == True,  # noqa
+            FileTable.c.supplementary == False
         )
+    )
+
+    logging.info('Setting availability of non-topic ContentNode objects based on File availability')
+
+    connection.execute(ContentNodeTable.update().where(
+        ContentNodeTable.c.id.in_(contentnode_statement)).values(available=True).execution_options(autocommit=True))
```

Comment:
Don't think it needs to be a blocker for merge, but perhaps making a list of tech debt / areas for further exploration, that we can look at post-merge?